Fixes #29219 - Deploy Storybook to surge.sh

jeremylenz opened the pull request:
Github secrets (and some other features) are not available for Github `pull_request` actions from PR's in forked repos.  So, this action runs instead on the `push` event on the `develop` branch.

The React storybook is deployed from the `develop` branch to https://foreman.surge.sh.

Requires the secrets `SURGE_TOKEN` and `SURGE_DOMAIN` (foreman.surge.sh) to be set on Github (Settings tab -> Secrets).  Secrets are properly stored and are not hard-coded.

You can check the status in the Actions tab.  To test the PR on a fork, comment out line 14:
```yaml
if: github.repository == 'theforeman/foreman'
```

jeremylenz committed:
Refs #29219 - put name key first

jeremylenz committed:
Fixes #29219 - Deploy Storybook to surge.sh
  Github pull_request events are not available for Github actions
  from PR's in forked repos.  So, the action is configured to run
  on a schedule daily at 7AM.

jeremylenz committed:
Refs #29219 - use yg/deploy-surge

jeremylenz committed:
Refs #29219 - address comments

jeremylenz commented:
See also @johnpmitsch [previous comments](https://github.com/theforeman/foreman/pull/7479#issuecomment-619078278) for more context on why this can't run per-PR

ekohl commented on .github/workflows/storybook_deploy_main.yml:
```diff
@@ -0,0 +1,28 @@
+name: Deploy Main Storybook
+
+on:
+  schedule:
+    - cron: '0 7 * * *'
```
Any reason to do this daily and not after a push?

```suggestion
on:
  push:
    branches:
      - develop
```

This will probably create an issue when we branch and the file ends up in multiple branches. It also prevents from every fork attempting to push to surge.sh.

jeremylenz replied:
@ekohl I don't think a PR merge triggers a push event.

jeremylenz replied:
Per [docs](https://help.github.com/en/actions/reference/events-that-trigger-workflows#scheduled-events-schedule), "Scheduled workflows run on the latest commit on the default or base branch."  I think that should be okay..?

ekohl replied:
Pretty sure it does. See https://github.com/theforeman/puppet-katello_devel/actions which has builds for master, which were certainly not PRs.

johnpmitsch replied:
My original thought was to run the action when:
- `develop` is pushed to
- a storybook-related directory was updated

So this will avoid deploying on every push to develop, the majority of which are not storybook updates.

This seems possible with [the 'Example including paths' section here](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths)

You could test this out on your fork if you are not already.

ekohl commented on .github/workflows/storybook_deploy_main.yml:
```diff
@@ -0,0 +1,28 @@
+      run: npm install
+    - name: Build Storybook
+      run: npm run build-storybook -- --output-dir=storybooks/main
+    - name: Deploy Storybook to surge.sh
+      id: surge_deploy
+      env:
+        SURGE_LOGIN: ${{ secrets.SURGE_LOGIN }}
+        SURGE_TOKEN: ${{ secrets.SURGE_TOKEN }}
+      if: ${{ env.SURGE_TOKEN }}
+      run: |
+        npm install --save-dev surge
+        node_modules/.bin/surge --project storybooks/main --domain foreman.surge.sh
```
Would it make sense to use an existing action for this, like https://github.com/marketplace/actions/deploy-surge

johnpmitsch replied:
Can we use `npx surge` instead of `node_modules/.bin/surge`? [More info is here](https://blog.npmjs.org/post/162869356040/introducing-npx-an-npm-package-runner) if you are not familiar, its a binary that comes with npm and allows one to use local package binaries.

ekohl commented on .github/workflows/storybook_deploy_main.yml:
```diff
@@ -0,0 +1,28 @@
+
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        clean: false
```
What's the benefit of this?

jeremylenz replied:
This was leftover from previous iterations, it's probably not needed anymore.  Will remove

jeremylenz committed:
Refs #29219 - use push event with branches & file filter
Co-authored-by: Ewoud Kohl van Wijngaarden <ewoud@kohlvanwijngaarden.nl>
add paths
remove clean: false

jeremylenz commented:
@ekohl @johnpmitsch Thanks for the feedback!

* Updated to run on `push` events instead of schedule.  I'm glad I was wrong about that.
* Updated to use `yg/deploy-surge@v1.0.1` action (so no need for `npx`, but will keep that in mind, thanks!)

ekohl commented on .github/workflows/storybook_deploy_main.yml:
```diff
@@ -0,0 +1,27 @@
+      name: Deploy Storybook to surge.sh
+      with:
+        SURGE_TOKEN: ${{ secrets.SURGE_TOKEN }}
+        DOMAIN_NAME: foreman.surge.sh
```
I'd also keep this as a secret tbh. This can allow people to easily set their own domain in a fork to show their own storybook if they want to.

ekohl commented on .github/workflows/storybook_deploy_main.yml:
```diff
@@ -0,0 +1,27 @@
+    branches:
+      - develop
+    paths:
+      - '**[Ss]tories**'
```
Isn't it just `webpack/stories` where we store everything?

jeremylenz replied:
`<componentName>.stories.js` files are usually stored next to the component, so they can be in lots of places.

ekohl replied:
Wouldn't it be sufficient to just list `webpack` then? I think the storybook can change if components change, even if the story itself doesn't change.

ekohl reviewed (commented):
Something I still haven't figured out with Github actions is how you prevent it from running on forks.

johnpmitsch commented:
> Something I still haven't figured out with Github actions is how you prevent it from running on forks.

I imagine the the secrets would not carry over to forks and prevent anything from actually deploying. But if we want to be thorough, it looks like you can use an `if` conditional to check the repo. https://github.community/t5/GitHub-Actions/Stop-github-actions-running-on-a-fork/td-p/51499

Something like:
`if: github.repository == 'theforeman/foreman'`

jeremylenz commented:
Updated to address comments, and also updated PR description

ekohl commented on .github/workflows/storybook_deploy_main.yml:
```diff
@@ -0,0 +1,28 @@
+      run: npm install
+    - name: Build Storybook
+      run: npm run build-storybook -- --output-dir=storybooks/main
+    - uses: yg/deploy-surge@v1.0.1
+      name: Deploy Storybook to surge.sh
```
Nit: I'd keep the name before uses.

amirfefer commented:
Thanks @jeremylenz ! a great addition :+1: 
does it make sense to use a more  specific name for the domain?  i.e `foreman-storybook.surge.sh` ?

jeremylenz commented:
@amirfefer I'd be fine with that :+1:

sharvit commented:
Thanks @jeremylenz looks good!

1. Assuming I want to use it in my plugin, what would I have to do?
Is there a way to share and reuse it?

2. How can we test it is actually work before merging?

sharvit commented:
> does it make sense to use a more specific name for the domain? i.e `foreman-storybook.surge.sh` ?

We can also assign an official subdomain like storybook.theforeman.org

jeremylenz commented:
> 2\. How can we test it is actually work before merging?

It's kind of a pain, but here's what I was doing:

1. Your fork of foreman will become the 'base' branch for purposes of testing
2. Use a different github account that you have (or make one) and fork YOUR foreman repo (not theforeman/foreman) so you can submit a PR from a forked repo to your 'base' branch.
3. Log out of Github and switch back to your 'base' branch. In your 'base' branch, set the Github secrets `SURGE_DOMAIN` and `SURGE_TOKEN`.
4. Submit a PR from the throwaway Github account to your 'base' branch.  (MAKE SURE you don't accidentally choose theforeman/foreman as the base)  Make sure it changes one of the files in `webpack/` somewhere
5. Check the progress of the job on the Actions tab.

You can also comment out `branches: develop` and just push directly, if you want to test the action runner without testing the full workflow.  That will at least demonstrate that the Github secrets and surge.sh deployment are working.

You can check out the [pull requests](https://github.com/jeremylenz/foreman/pulls?q=is%3Apr+is%3Aclosed) on my branch if you want to see all the ugliness of my testing.

also, here's one of my successful [actions](https://github.com/jeremylenz/foreman/runs/667653318?check_suite_focus=true).

ekohl commented:
> * Your fork of foreman will become the 'base' branch for purposes of testing
> 
> * Use a different github account that you have (or make one) and fork YOUR foreman repo (not theforeman/foreman) so you can submit a PR from a forked repo to your 'base' branch.
> 
> * Log out of Github and switch back to your 'base' branch. In your 'base' branch, set the Github secrets `SURGE_DOMAIN` and `SURGE_TOKEN`.
> 
> * Submit a PR from the throwaway Github account to your 'base' branch.  (MAKE SURE you don't accidentally choose theforeman/foreman as the base)  Make sure it changes one of the files in `webpack/` somewhere

This feels *way* too complicated. You can actually submit PRs to your own repository, regardless of whether you forked `theforeman/foreman` or some other repo.

Edit: since this runs after a new commit has been pushed, you can also just push straight to your base branch.

jeremylenz commented:
> 1. Assuming I want to use it in my plugin, what would I have to do?
>    Is there a way to share and reuse it?

I didn't really think about this.  This will only trigger on push events in theforeman/foreman repo.  Plugins could add a similar Github Actions workflow, I suppose, and it should work just fine.  But the domain should probably be different.

sharvit commented:
> > 1. Assuming I want to use it in my plugin, what would I have to do?
> >    Is there a way to share and reuse it?
> 
> I didn't really think about this. This will only trigger on push events in theforeman/foreman repo. Plugins could add a similar Github Actions workflow, I suppose, and it should work just fine. But the domain should probably be different.

So I guess you can merge it to you `develop` branch (or another branch it doesn't matter IMO) and share the log with us as it is public?

@tbrisker, can you please add the `SURGE_TOKEN` and `SURGE_DOMAIN` to the github environment variables?

sharvit commented:
Looks good to me, will approve once we will have the secret keys available so it would actually work.

jeremylenz commented:
@tbrisker emailed you the keys :)

tbrisker commented:
Secrets have been added to the repo.
@ekohl any other concerns from your side?

ekohl commented:
@tbrisker I don't see any credentials in our infra shared secret storage.

tbrisker commented:
> @tbrisker I don't see any credentials in our infra shared secret storage.

Good point, forgot to add them there, i'll push soon.

tbrisker commented:
@ekohl pushed.

ekohl commented:
In gopass I only see a domain and token, but no full account details. This means as the infra team we have no way to manage the account. If the token is expired, we can't fix it. Because we've had too many of these issues in the past (whenever a person moved on different things), I'd like to get that sorted before we start using it.

jeremylenz commented:
@ekohl okay, I sent the account info to @tbrisker.

tbrisker commented:
Pushed to the store.

tbrisker reviewed (approved):
Thanks @jeremylenz and all!

tbrisker reviewed (changes requested):
Or actually, one small change please - can you update the README.md to point to surge instead of the (outdated) gh-pages?

jeremylenz committed:
Refs #29129 - Update README

tbrisker reviewed (approved):
Test failures irrelevant. Thanks @jeremylenz !
